### PR TITLE
[Core] Add dedicated DocStringTypeRegistry

### PR DESCRIPTION
**Pull request by rasklaad**
## Summary

Moved docstring transformation from 1x1 DataTable to separate class

## Motivation and Context

https://github.com/cucumber/cucumber-jvm/issues/1705

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue).
- [ ] New feature (non-breaking change which adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected).

## Checklist:

- [x] I've added tests for my code.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: core/src/main/java/io/cucumber/core/stepexpression/DocStringType.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
+package io.cucumber.core.stepexpression;
+
+
+import java.lang.reflect.Type;
+
+public class DocStringType {
```
Comment (mpkorstanje) [inline core/src/main/java/io/cucumber/core/stepexpression/DocStringType.java]:
```suggestion
@API(status = API.Status.STABLE)
public final class DocStringType {
```

And you may have to fix the import manually.

File: core/src/main/java/io/cucumber/core/stepexpression/DocStringType.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    private final Type targetType;
+    private final String contentType;
+    private final Converter converter;
+
+    public Converter getConverter() {
```
Comment (mpkorstanje) [inline core/src/main/java/io/cucumber/core/stepexpression/DocStringType.java]:
```suggestion
    Converter getConverter() {
```

File: core/src/main/java/io/cucumber/core/stepexpression/DocStringType.java [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+        return converter;
+    }
+
+    public String getContentType() {
```
Comment (mpkorstanje) [inline core/src/main/java/io/cucumber/core/stepexpression/DocStringType.java]:
```suggestion
    String getContentType() {
```

File: core/src/main/java/io/cucumber/core/stepexpression/DocStringType.java [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+        return contentType;
+    }
+
+    public Type getTargetType() {
```
Comment (mpkorstanje) [inline core/src/main/java/io/cucumber/core/stepexpression/DocStringType.java]:
```suggestion
    Type getTargetType() {
```

File: core/src/main/java/io/cucumber/core/stepexpression/DocStringType.java [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+        if (contentType == null) {
+            this.contentType = "";
+        }
+        else {
+            this.contentType = contentType;
+        }
+        this.converter = converter;
+    }
+
+    public interface Converter<T> {
```
Comment (mpkorstanje) [inline core/src/main/java/io/cucumber/core/stepexpression/DocStringType.java]:
```suggestion
    @FunctionalInterface
    public interface Converter<T> {
```

**Comment by rasklaad** [comment]
Hi @mpkorstanje , thank you for the review.
I almost fixed everything you pointed, and I have a question how DocStringConverter should behave in this situation:
Feature:
```
Given doc string step
   """my-content-type
   really long docstring
   """
```
Step def:
```
@Given("doc string step")
public void step(String docString) {
    System.out.println(docString);
}
```
I have a two options:
1. DocStringConverter throw exception, because `my-content-type` is not registered.
 This will probably broke a lot of the user code, if user use content type without registering custom converter.
2. DocStringConverter ignore content-type, because target type is String.
This will not allow user register something like this:
```
registry.defineDocStringType(new DocStringType(String.class, "my-content-type", (String s) -> s.replaceAll("\\s", "-")))
```

**Comment by mpkorstanje** [comment]
I think that by searching for possible converter in order of priority of name, then class, should solve that problem.

```
If a doc string type is provided in the gherkin lookup a converter for that doc string type
   if a converter for that doc string type could be found
        convert the value using the the found converter
        return the converted value

if a converter for the type could be found
  convert the value using the the found converter
  return the converted value

throw exception no converter could be found by doc string type, type and no default converter was available.
```

File: core/src/test/java/io/cucumber/core/stepexpression/StepExpressionFactoryTest.java [hunk 6]
```diff
@@ -103,13 +104,39 @@ void unknown_target_type_does_no_transform_data_table() {
...
+        StepExpression expression = new StepExpressionFactory(registry).createExpression("Given some stuff:", JsonNode.class);
+        List<Argument> match = expression.match("Given some stuff:", docString, contentType);
+        JsonNode node = (JsonNode) match.get(0).getValue();
+        assertThat(node.asText(), equalTo(docString));
+    }
+
+    @Test
+    void docstring_expression_transform_anonymous_doc_string_type() {
+        String docString = "A rather long and boring string of documentation";
+        String contentType = null;
```
Comment (mpkorstanje) [inline core/src/test/java/io/cucumber/core/stepexpression/StepExpressionFactoryTest.java]: This should result in an error. The content type must always be a string. And the string may not be empty. With the exception of the nameless doc string type we predefined.

**Comment by mpkorstanje** [comment]
Added a few bits to make all these variations work:

```java
    @Given("a doc string 1")
    public void docstr(Object d) {

    }

    @Given("a doc string 2")
    public void docstr(DocString d) {
        d.convert(JsonNode.class);
    }

    @Given("a doc string 3")
    public void docstr(String d) {

    }

    @Given("a doc string 4")
    public void docstr(JsonNode d) {

    }
```

**Comment by mpkorstanje** [comment]
It's currently possible to define parameter types by annotation:

```
public class Steps {

    @Given("today is {iso8601Date}")
    public void today_is(LocalDate date) {

    }
}
```

We should make the same possible for doc strings. I've added the annotation but it's not added to the glue yet. I'm out of time for now. Feel free to pick that up if you have some time.

**Comment by rasklaad** [comment]
> It's currently possible to define data table types by annotation:
> 
> ```
>    @DataTableType
>     public Author authorEntryTransformer(Map<String, String> entry) {
>         return new Author(
>             entry.get("firstName"),
>             entry.get("lastName"),
>             entry.get("birthDate"));
>     }
>     
> ```
> 
> We should make the same possible for doc strings. I've added the annotation but it's not added to the glue yet. I'm out of time for now. Feel free to pick that up if you have some time.

Sure, I'll do it, but it will take some time, to figure out how @DataTableType works.

**Comment by mpkorstanje** [comment]
Good enough I think. Cheers!

**Pull request merged**